### ISSUE TITLE
GHCS bug in simple program

### ISSUE THREAD

**Issue by bbarker** [google/codeworld#724]
**Describe the bug**

This appears to be a GHCJS bug, but I'm new here:

```
panic! (the 'impossible' happened)
  (GHCJS version 0.2.1, GHC version 8.0.2):
	initTc: unsolved constraints
  WC {wc_insol =
        [W] fromString_a2MX :: t_a2MW[tau:1] (CHoleCan: fromString)
        [W] concatenation_a2N6 :: t_a2N5[tau:1] (CHoleCan: concatenation)
        [W] toString_a2Ng :: t_a2Nf[tau:1] (CHoleCan: toString)
        [W] toString_a2Np :: t_a2No[tau:1] (CHoleCan: toString)}

Please report this as a GHCJS bug:  https://github.com/ghcjs/ghcjs/issues

```

**Shared Program**

If you have a program that demonstrates what is wrong, add a link here from the "Share" button:
- https://code.world/haskell#PpuTGVco91DpCgxUkW_DqkQ

**Details**

- What browser are you using? Firefox/Vivaldi
- What operating system are you using? Ubuntu 18.04

**Additional context**

This was a fairly quick port from regular code world code; maybe I overlooked something

**Comment by cdsmith** [comment]
Minimal test case: https://code.world/haskell#P-PZMKwheSmtwIN0SjWd_Iw

    x :: Double -> Double
    x = y

**Comment by bbarker** [comment]
Is it a problem in vanilla ghcjs or just in code world?

**Comment by cdsmith** [comment]
CodeWorld is just invoking GHCJS on the server, so this is a GHCJS problem.  It's a weird one, though.

**Comment by cdsmith** [comment]
This is a GHC bug: https://ghc.haskell.org/trac/ghc/ticket/12921

There are two options here: update to a newer GHC (see #655), or backport the fix to 8.0 and patch the local GHC.  This problem only occurs when there's an out-of-scope identifier, so it never affects correct code.  I've also never run into it before, so despite being easy to reproduce with trivial code, it doesn't look like it's common in practice.  So I'm tempted to decide on waiting for upgrade to 8.2 or 8.4.

**Comment by cdsmith** [comment]
The patch is non-trivial, and I'd rather not try to fix it in 8.0.  @luite : Do you have any recommendation on the safety of migrating to GHCJS 8.2 or 8.4?

**Comment by cdsmith** [comment]
To quickly follow up, it looks like three things are necessary to trigger this bug.

1. There must be no definition of `main` (or `program` in CodeWorld mode).
2. There must be an explicit type signature on a declaration.
3. The declaration must use a variable that's not in scope.

The first and third are both independent errors, so it requires two separate errors to trigger this bug.  Nice to fix, but not a big deal.  I'll close this as a duplicate of #655.
